src/resolveElements.js

```diff
@@ -0,0 +1,104 @@
+import isPromise from 'is-promise';
+import React from 'react';
+import warning from 'warning';
+
+import HttpError from './HttpError';
+
+const UNRESOLVED = {};
+
+function checkResolved(value) {
+  if (!isPromise(value)) {
+    return value;
+  }
+
+  return Promise.race([value, UNRESOLVED]);
```

Comment:
its clever, tho it seems like it could be wrong if promise is `Promise.Resolve(true)` i have no idea why that would resolve out of order but maybe it can b/c Javascript? who knows